catkin_ws/src/state_estimation/src/ekf.py

```diff
@@ -0,0 +1,391 @@
+from aquadrone_msgs.msg import SubState, MotorControls
+
+from ekf_indices import IDx as IDx
+from ekf_sensors import IMUSensorListener, PressureSensorListener
+
+def quat_msg_to_vec(q):
+    return np.array([q.w, q.x, q.y, q.z])
+
+def rpy_vec_to_msg(vec):
+    angles = Vector3()
+    angles.x = vec[0]
+    angles.y = vec[1]
+    angles.z = vec[2]
+    return angles
+
+def msg_quaternion_to_euler(quat):
+    q_vec = quat_msg_to_vec(quat)
+    rpy_vec = quaternion_to_euler(q_vec)
+    return rpy_vec_to_msg(rpy_vec)
+
+def quaternion_to_euler(quat_vec):
+        # From wikipedia (https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles)
+
+        w = quat_vec[0]
+        x = quat_vec[1]
+        y = quat_vec[2]
+        z = quat_vec[3]
+
+        angles = np.array([0, 0 ,0])
+
+        # roll (x-axis rotation)
+        sinr_cosp = +2.0 * (w * x + y * z)
+        cosr_cosp = +1.0 - 2.0 * (x * x + y * y)
+        #angles[0] = np.arctan2(sinr_cosp, cosr_cosp)
+        
+
+        # yaw (z-axis rotation)
+        siny_cosp = +2.0 * (w * z + x * y)
+        cosy_cosp = +1.0 - 2.0 * (y * y + z * z)
+        #angles[2] = math.atan2(siny_cosp, cosy_cosp)
+
+        # pitch (y-axis rotation)
+        sinp = +2.0 * (w * y - z * x)
+        if np.abs(sinp) >= 1:
+            #angles[1] = np.sign(sinp) * math.pi # use 90 degrees if out of range
+            return np.array([ np.arctan2(sinr_cosp, cosr_cosp),
+                              np.sign(sinp) * math.pi,
+                              np.arctan2(siny_cosp, cosy_cosp)
+                              ])
+        else:
+            #angles[1] = math.asin(sinp)
+            return np.array([ np.arctan2(sinr_cosp, cosr_cosp),
+                              np.arcsin(sinp),
+                              np.arctan2(siny_cosp, cosy_cosp)
+                              ])
+
+        return angles
+
+
+quat_to_euler_jacobian = jacobian(quaternion_to_euler)
+
+
+
+class EKF:
+    def __init__(self, config):
+        # https://en.wikipedia.org/wiki/Extended_Kalman_filter
+
+        ''' Model Description
+        x = state
+        u = inputs (thruster forces, etc)
+        z = ouputs (measurements)
+        P = unvertainty/variance matrix of state x
+
+        Dynamics: x[k+1] = f(x[k], u[k])
+        Outputs:    z[k] = h(x[k], u[k])
+
+        Linear Form:
+          x[k+1] = A*x[k] + B*u[k]
+            z[k] = C*x[k] + B*u[k]
+        '''
+
+        self.n = 13 # Number of state elements
+        self.m = 8 # Number of inputs
+
+        self.x = np.zeros((self.n, 1))
+        self.x[IDx.Ow] = 1
+
+        self.u = np.zeros((self.m,1))
+
+        self.B = np.array(config.get_thrusts_to_wrench_matrix())
+
+        self.P = np.eye(self.n)
+        self.Q = np.eye(self.n) * 0.01 # Uncertanty in dynamics model
+
+        self.depth_sub = PressureSensorListener()
+        self.imu_sub = IMUSensorListener()
+
+        self.rate = 20
```

Comment:
Hey Wesley, so just to summarize the goal is to run the EKF loop at 20Hz (or another rate based on testing).

Each loop will do a prediction step based on the motor thrusts, gravity, bouyancy, drag, and other forces to update the expected state and its variance. Then a measurement step, where it uses input from the pressure sensor and gyro (and eventually the zed mini and other sensors as well) to refine its expected state and variance. Then there is a step where the new expected state/variance is converted to a SubState message which is then published.

The assumption here is that this entire loop will be able to run within 20Hz every single time (because otherwise the time step being used in calculations would be wrong).

Reply:
That's a good summary; I might add that as a comment in the code for others.

The timing could be a potential issue in the way you mention. Tracking something like the last time it was run might be better for it.

Reply:
I've changed how dt is calculated